src/weights.jl

```diff
@@ -2,44 +2,191 @@
 else
-    immutable WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
-        values::V
-        sum::S
+    @compat abstract type AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T} end
+end
+
+"""
+    @weights name
+
+Generates a new generic weight type with specified `name`, which subtypes `AbstractWeights`
+and stores the `values` (`V<:RealVector`) and `sum` (`S<:Real`).
+"""
+macro weights(name)
+    return quote
+        if VERSION < v"0.6.0-dev.2123"
+            immutable $name{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        else
+            immutable $name{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        end
     end
 end
 
+eltype(wv::AbstractWeights) = eltype(wv.values)
+length(wv::AbstractWeights) = length(wv.values)
+values(wv::AbstractWeights) = wv.values
+sum(wv::AbstractWeights) = wv.sum
+isempty(wv::AbstractWeights) = isempty(wv.values)
+
+Base.getindex(wv::AbstractWeights, i) = getindex(wv.values, i)
+Base.size(wv::AbstractWeights) = size(wv.values)
+
+"""
+    varcorrection(n::Integer, corrected=false)
+
+Compute a bias correction factor for calculating `var`, `std` and `cov` with
+`n` observations. Returns ``\\frac{1}{n - 1}`` when `corrected=true`
+(i.e. [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
+otherwise returns ``\\frac{1}{n}`` (i.e no correction).
+"""
+varcorrection(n::Integer, corrected::Bool=false) = 1 / (n - Int(corrected))
+
+
+@weights AnalyticWeights
+
 """
-    WeightVec(vs, [wsum])
+    AnalyticWeights(vs, wsum=sum(vs))
+
+Construct an `AnalyticWeights` vector with weight values `vs`.
+A precomputed sum may be provided as `wsum`.
 
-Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
-If omitted, `wsum` is computed.
+Analytic weights describe a non-random relative importance (usually between 0 and 1)
+for each observation. These weights may also be referred to as reliability weights,
+precision weights or inverse variance weights.
 """
-function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
-    return WeightVec{S, eltype(vs), V}(vs, s)
+AnalyticWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs)) =
+    AnalyticWeights{S, eltype(vs), V}(vs, s)
+
+"""
+    aweights(vs)
+
+Construct an `AnalyticWeights` vector from array `vs`.
+See the documentation for `AnalyticWeights` for more details.
+"""
+aweights(vs::RealVector) = AnalyticWeights(vs)
+aweights(vs::RealArray) = AnalyticWeights(vec(vs))
+
+"""
+    varcorrection(w::AnalyticWeights, corrected=false)
+
+``\\frac{1}{\sum w - \sum {w^2} / \sum w}``
+"""
+function varcorrection(w::AnalyticWeights, corrected::Bool=false)
+    s = w.sum
+
+    if corrected
+        sum_sn = zero(eltype(w)) / one(typeof(s)) ^ 2  # to ensure type stability
```

Comment:
The second benchmark is wrong, since it doesn't return `sum_sn` the loop is optimized out. Wrapping everything in functions to make sure the closure is dealt with properly, I get this:
```julia
julia> function f1(w, s)
          sum_sn = zero((zero(eltype(w)) / one(typeof(s))) ^ 2)  # to ensure type stability
          @inbounds for x in w
              sum_sn += (x / s) ^ 2
          end
          sum_sn
       end
f1 (generic function with 1 method)

julia> f2(w, s) = sum(x -> (x / s) ^ 2, w)
f2 (generic function with 1 method)

julia> w = collect(1:100_000);

julia> @benchmark f1($w, $(sum(w)))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     868.741 μs (0.00% GC)
  median time:      870.310 μs (0.00% GC)
  mean time:        894.750 μs (0.00% GC)
  maximum time:     1.217 ms (0.00% GC)
  --------------
  samples:          5580
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark f2($w, $(sum(w)))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     97.496 μs (0.00% GC)
  median time:      99.200 μs (0.00% GC)
  mean time:        103.157 μs (0.00% GC)
  maximum time:     380.056 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

```

I'm not sure why `sum` is faster in this case...